curvefs/src/mds/schedule/copysetSchedule.cpp

```diff
@@ -0,0 +1,267 @@
+    LOG(INFO) << "copyset scheduler gen " << op.OpToString() << " on "
+              << copyset.CopySetInfoStr();
+
+    // add operator
+    if (!opController_->AddOperator(op)) {
+        LOG(INFO) << "copysetSchduler add op " << op.OpToString()
+                  << " fail, copyset has already has operator"
+                  << " or operator num exceeds the limit.";
+        return false;
+    }
+
+    // create copyset
+    if (!topo_->CreateCopySetAtMetaServer(copyset.id, destId)) {
+        LOG(ERROR) << "copysetScheduler create " << copyset.CopySetInfoStr()
+                   << " on metaServer: " << destId << " error, delete operator"
+                   << op.OpToString();
+        opController_->RemoveOperator(copyset.id);
+        return false;
+    }
+
+    LOG(INFO) << "copysetScheduler create " << copyset.CopySetInfoStr()
+              << "on metaServer:" << destId
+              << " success. generator op: " << op.OpToString() << "success";
+    return true;
+}
+
+bool CopySetScheduler::IsCopysetCanTransfer(const CopySetInfo &copyset) {
+    // return false if operator exists on copyset
+    Operator exist;
+    if (opController_->GetOperatorById(copyset.id, &exist)) {
+        return false;
+    }
+
+    // return false if copyset already has candidate
+    if (copyset.HasCandidate()) {
+        LOG(WARNING) << copyset.CopySetInfoStr() << " already has candidate: "
+                     << copyset.candidatePeerInfo.id;
+        return false;
+    }
+
+    // return false if the replica num of copyset is not standard
+    uint16_t replicaNum = topo_->GetStandardReplicaNumInPool(copyset.id.first);
+    if (copyset.peers.size() != replicaNum) {
+        LOG(WARNING) << copyset.CopySetInfoStr()
+                     << " peer num not match pool replica num: " << replicaNum;
+        return false;
+    }
+
+    // some peers are offline
+    if (!CopysetAllPeersOnline(copyset)) {
+        LOG(WARNING) << copyset.CopySetInfoStr() << " some peers are offline";
+        return false;
+    }
+
+    return true;
+}
+
+int CopySetScheduler::CopySetScheduleOverloadMetaserver(PoolIdType poolId) {
+    auto copysetList = topo_->GetCopySetInfosInPool(poolId);
+    auto metaserverList = topo_->GetMetaServersInPool(poolId);
+
+    std::map<MetaServerIdType, std::vector<CopySetInfo>> distribute;
+    GetCopySetDistribution(copysetList, metaserverList, &distribute);
+
+    int genOpCount = 0;
+    for (const auto &ms : metaserverList) {
+        // Find if any metaserver which resource exceeds the threshold,
+        // skip unhealthy metaserver
+        if (!ms.IsHealthy() || !ms.IsResourceOverload()) {
+            continue;
+        }
+
+        MetaServerIdType sourceId = ms.info.id;
+        LOG(INFO) << "metaserver resource is overload, should transfer some "
+                  << "copyset to other metaserver, metaserverId = " << sourceId;
+
+        for (const auto &copyset : distribute[sourceId]) {
```

Comment:
if one metaserver is overloaded, you just randomly select one copyset and transfer it into another metaserver.

should you find one or some copysets that can lead to minimizing migration effort

Reply:
With the current version, I don't know how to measure the copy migration cost. I can't get how much resouce the copyset used.
I can add a todo here.